### PR TITLE
Add NumPy RNG constructor `Op`s

### LINKED ISSUE #739: Create a `RandomState` `Variable` constructor `Op`
We need an `Op` that will construct `RandomState` `Variable`s with the seed values as inputs.  

We should need only one such `Op` that handles all `RandomState` types (e.g. by having a fixed `Op`-level property that determines which types the `Op` generates).

This&mdash;combined with #738&mdash;will largely remove the need to use shared variables for RNG states.

### PR DESCRIPTION
**Pull request by kc611**
Resolves #739 

Adds `RandomStateMakerOp` and `DefaultGeneratorMakerOp` as constructor `Op`s for `RandomType` `Variable`s

### CODE REVIEW

**Comment by kc611** [comment]
Tried adding a custom `RandomStateVariable` for these `Op`s, ran into couple of problems during constant folding and other few places.

File: aesara/tensor/random/op.py [hunk 1]
```diff
@@ -399,3 +399,43 @@ def grad(self, inputs, outputs):
 
     def R_op(self, inputs, eval_points):
         return [None for i in eval_points]
+
+
+class RandomMakerOp(Op):
+    def make_node(self, seed=None):
+        # np.inf is used as an representative of seed = None
+        if seed is None:
+            seed = np.inf
```
Comment (brandonwillard) [inline aesara/tensor/random/op.py:408]: `None` can be used in a graph (see `aesara.tensor.type_other.NoneConst`), and we should use it here.

File: aesara/tensor/random/op.py [hunk 2]
```diff
@@ -399,3 +399,43 @@ def grad(self, inputs, outputs):
...
+        inputs = [seed]
+        outputs = [self.random_type()]
+        return Apply(self, inputs, outputs)
+
+    def perform(self, node, inputs, output_storage):
+        (seed,) = inputs
+        # np.inf is used as an representative of seed = None
+        if seed == np.inf:
+            seed = None
+        output_storage[0][0] = self.bitgen(seed=seed)
```
Comment (brandonwillard) [inline aesara/tensor/random/op.py:419]: This implies that there should be an abstract method named `bitgen` in the interface defined by `RandomMakerOp`.  It can be added using [`abc.abstractmethod`](https://docs.python.org/3/library/abc.html#abc.abstractmethod).

However, it doesn't look like an abstract implementation of `Op.perform` is necessary.  When `np.inf` is replaced by `None` (see above), there's no code duplicated between subclasses, because only a direct call the underlying RNG constructor is required.

File: aesara/tensor/random/op.py [hunk 3]
```diff
@@ -399,3 +399,43 @@ def grad(self, inputs, outputs):
 
     def R_op(self, inputs, eval_points):
         return [None for i in eval_points]
+
+
+class RandomMakerOp(Op):
```
Comment (brandonwillard) [inline aesara/tensor/random/op.py:404]:
```suggestion
class AbstractRNGConstructor(Op):
```

File: aesara/tensor/random/op.py [hunk 4]
```diff
@@ -399,3 +399,43 @@ def grad(self, inputs, outputs):
...
+
+    def perform(self, node, inputs, output_storage):
+        (seed,) = inputs
+        # np.inf is used as an representative of seed = None
+        if seed == np.inf:
+            seed = None
+        output_storage[0][0] = self.bitgen(seed=seed)
+
+
+class RandomStateMakerOp(RandomMakerOp):
```
Comment (brandonwillard) [inline aesara/tensor/random/op.py:422]:
```suggestion
class RandomStateConstructor(RandomMakerOp):
```

File: aesara/tensor/random/op.py [hunk 5]
```diff
@@ -399,3 +399,43 @@ def grad(self, inputs, outputs):
...
+    random_type = RandomStateType()
+
+    def bitgen(self, seed):
+        return np.random.RandomState(seed=seed)
+
+
+random_state_maker = RandomStateMakerOp()
```
Comment (brandonwillard) [inline aesara/tensor/random/op.py:429]:
```suggestion
RandomState = RandomStateConstructor()
```

This mirrors the NumPy interface and needs to be exposed via `aesara.tensor.random.__init__`.

File: aesara/tensor/random/op.py [hunk 6]
```diff
@@ -399,3 +399,43 @@ def grad(self, inputs, outputs):
...
+class DefaultGeneratorMakerOp(RandomMakerOp):
+    random_type = RandomGeneratorType()
+
+    def bitgen(self, seed):
+        if seed.size == 1:
+            seed = int(seed)
+        return np.random.default_rng(seed=seed)
+
+
+default_rng_maker = DefaultGeneratorMakerOp()
```
Comment (brandonwillard) [inline aesara/tensor/random/op.py:441]:
```suggestion
default_rng = DefaultGeneratorMakerOp()
```

This mirrors the NumPy interface and needs to be exposed via `aesara.tensor.random.__init__`.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/tensor/random/op.py (hunk 2 above)
Comment (kc611) [reply aesara/tensor/random/op.py]: I made `bitgen` rapper function around those constructors primarily because directly having those constructors as class methods passed `self` into them by default. But I think the current approach works as well.

File: aesara/tensor/random/op.py [hunk 7]
```diff
@@ -399,3 +400,36 @@ def grad(self, inputs, outputs):
...
+        else:
+            seed = as_tensor_variable(seed)
+        inputs = [seed]
+        outputs = [self.random_type()]
+        return Apply(self, inputs, outputs)
+
+    def perform(self, node, inputs, output_storage):
+        (seed,) = inputs
+        if seed is not None and seed.size == 1:
+            seed = int(seed)
```
Comment (kc611) [inline aesara/tensor/random/op.py:418]: Had to do this because of what seems to be a bug in Numpy

```python
import numpy as np

x = np.random.default_rng(3) #Works
x = np.random.default_rng(np.array([3])) #Works
x = np.random.default_rng(np.array(3)) #Throws error
```

**Review by brandonwillard — CHANGES_REQUESTED** [review]
This looks great!  We just need a test for `seed=None` and we can merge.
